homework/Truong_Duong/lesson_4/javascript_exercise.md

```diff
@@ -0,0 +1,203 @@
+> For the second code block, the iterator variable `i` is defined with `let`, it have block scope. Therefore, when `setTimeout` function reference to variable `i`, due to `closure` in JS, the value of `i` is wrap to the execution context of each `setTimeout` function. Result: `0 1 2`
+
+
+**3. What's the output?**
+
+```js
+const shape = {
+  radius: 10,
+  diameter() {
+    return this.radius * 2;
+  },
+  perimeter: () => 2 * Math.PI * this.radius,
+};
+ 
+console.log(shape.diameter());
+console.log(shape.perimeter());
+```
+
+> **Answer: 20, NaN**
+> The main point is the value `this` in function declaration (`diameter`) vs arrow function (`perimeter`)
+> `this` in function declaration is `dynamic`, depend on the context => In this case: `this` is shape object
+> `this` in arrow function is `bound lexically`, it is equal to `this` value of the nearest outer scope, which is `global` object in this case. `global` have no radius attribute => `global.radius = undefined`, and `number * undefined = NaN`
+
+**4. What's the output?**
+
+```js
++true;
+!'Lydia';
+```
+
+A: 1 and false
+B: false and NaN
+C: false and false
+ 
+> Answer: A
+> By default, JS will try to convert the input to its expected type.
+> `+true` return 1. After `+` operator, JS expect a number, true is converted to 1
+> `!'Lydia'` return `false`. After `!` operator, JS expect a boolean, it compare `Lydia === true`, therefore `Lydia` is treated as `true` and `!true = false`
+
+**5. Which one is true?**
+
+```js
+const bird = {
+  size: 'small',
+};
+ 
+const mouse = {
+  name: 'Mickey',
+  small: true,
+};
+```
+
+A: mouse.bird.size is not valid
+B: mouse[bird.size] is not valid
+C: mouse[bird["size"]] is not valid
+D: All of them are valid
+
+> **Answer: A**
+> Option A: `mouse.bird` return `undefined`, then access `size` of `undefined` return error
+> Option B, C: are almost similar, just different syntax. First JS will evaluate `bird.size` (`bird["size"]`), then return `small`, `mouse.small` return `true`.
+
+**6. What's the output?**
+
+```js
+let c = { greeting: 'Hey!' };
+let d;
+d = c;
+c.greeting = 'Hello';
+console.log(d.greeting);
+```
+
+A: Hello
+B: Hey!
+C: undefined
+D: ReferenceError
+E: TypeError
+ 
+> **Answer: A**
+> Object is reference type, meaning that when we assign `d=c`, `d` is the pointer to the area where c value is store. Therefore, `d.greeting` is equal to `c.greeting`
+
+**7. What's the output?**
+
+```js
+let a = 3;
+let b = new Number(3);
+let c = 3;
+console.log(a == b);
+console.log(a === b);
+console.log(b === c);
+```
+
+A: true false true
+B: false false true
+C: true false false
+D: false true true
+
+> **Answer: C**
+> Reference here: [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)
+> In short, `Number('123') === 123`
+> When Number is called as a constructor (with new), it creates a `Number object`, which is not a primitive. `typeof new Number(42) === "object"`, and `new Number(42) !== 42`
```

Comment:
Honestly, I have never use `new Number()` or `new Boolean(), new String(), ...`. I get used to using these classes for casting variables like `Number('3') or Boolean(-1)` :D
This information help me a lot. Thanks